pettingzoo/butterfly/knights_archers_zombies/knights_archers_zombies.py

```diff
@@ -385,129 +275,124 @@ def observe(self, agent):
+            self.arrow_hit()
 
             # Zombie Kills the Sword
-            self.zombie_list, self.sword_list, self.all_sprites, self.score = self.zombie_sword(self.zombie_list, self.sword_list, self.all_sprites, self.score)
+            self.sword_hit()
 
             # Zombie Kills the Archer
             if self.killable_archers:
-                self.zombie_archer(self.zombie_list, self.archer_list, self.all_sprites, self.archer_killed)
+                self.zombie_hit_archer()
 
             # Zombie Kills the Knight
             if self.killable_knights:
-                self.zombie_list, self.knight_list, self.all_sprites, self.knight_killed, self.sword_list, self.sword_killed = self.zombie_knight(self.zombie_list, self.knight_list, self.all_sprites, self.knight_killed, self.sword_list, self.sword_killed)
-
-            # Kill the Sword when Knight dies
-            self.sword_killed, self.sword_list, self.all_sprites = self.kill_sword(self.sword_killed, self.sword_list, self.all_sprites)
+                self.zombit_hit_knight()
 
+            # update some zombies
             for zombie in self.zombie_list:
                 zombie.update()
-            arrows_to_delete = []
 
+            # update some arrows
             for arrow in self.arrow_list:
                 arrow.update()
-                if not arrow.is_active():
-                    arrows_to_delete.append(arrow)
-            for arrow in arrows_to_delete:
-                self.arrow_list.remove(arrow)
-                self.all_sprites.remove(arrow)
+
+            # Spawning Zombies at Random Location at every 100 iterations
+            self.spawn_zombie()
 
             self.draw()
 
             self.check_game_end()
             self.frames += 1
 
         self._clear_rewards()
-        self.rewards[agent] = agent_name.score
-        agent_name.score = 0
+        self.rewards[self.agent_selection] = agent.score
+        agent.score = 0
         done = not self.run or self.frames >= self.max_cycles
-        self.dones = {a: done for a in self.agents}
+        self.dones = {a: done for a in self.possible_agents}
 
+        # manage the kill list
         if self._agent_selector.is_last():
-            _live_agents = self.agents[:]
+            for agent in self.kill_list:
+                self.agents.remove(agent)
+                self.dead_agents.append(agent)
 
-            for k in self.kill_list:
-                self.dones[k] = True
-                _live_agents.remove(k)
+            for agent in self.dead_agents:
+                self.dones[agent] = True
 
             # reset the kill list
             self.kill_list = []
 
-            self._agent_selector.reinit(_live_agents)
+            self._agent_selector.reinit(self.agents)
+
         if len(self._agent_selector.agent_order):
             self.agent_selection = self._agent_selector.next()
-        self._cumulative_rewards[agent] = 0
+
+        self._cumulative_rewards[self.agent_selection] = 0
```

Comment:
This change breaks the reward. The previous agent's reward should get set to 0, the self.agent_selection is the next agent at this point in the function.